src/dune_pkg/fetch.ml

```diff
@@ -260,6 +261,149 @@ let fetch ~unpack ~checksum ~target (url : OpamUrl.t) =
         url)
 ;;
 
+module Git_repo = struct
+  type t =
+    { git : Path.t
+    ; dir : Path.t
+    }
+
+  type rev = Rev of string
+
+  let equal { git; dir } t = Path.equal git t.git && Path.equal dir t.dir
+  let display = Display.Quiet
+  let failure_mode = Process.Failure_mode.Strict
+  let output_limit = Dune_engine.Execution_parameters.Action_output_limit.default
+  let make_stdout () = Process.Io.make_stdout ~output_on_success:Swallow ~output_limit
+  let make_stderr () = Process.Io.make_stderr ~output_on_success:Swallow ~output_limit
+
+  let run { git; dir } =
+    let stdout_to = make_stdout () in
+    let stderr_to = make_stderr () in
+    Process.run ~dir ~display ~stdout_to ~stderr_to failure_mode git
+  ;;
+
+  let run_capture_line { git; dir } =
+    Process.run_capture_line ~dir ~display failure_mode git
+  ;;
+
+  let run_capture_lines { git; dir } =
+    Process.run_capture_lines ~dir ~display failure_mode git
+  ;;
+
+  let run_capture_zero_separated_lines { git; dir } =
+    Process.run_capture_zero_separated ~dir ~display failure_mode git
+  ;;
+
+  let show { git; dir } (Rev rev) path =
+    let open Fiber.O in
+    let failure_mode = Dune_vcs.Vcs.git_accept () in
+    let command = [ "show"; sprintf "%s:%s" rev (Path.to_string path) ] in
+    let stderr_to = make_stderr () in
+    let+ res =
+      Process.run_capture_lines ~dir ~display ~stderr_to failure_mode git command
+    in
+    match res with
+    | Error _n -> None
+    | Ok v -> Some v
+  ;;
+
+  let load ~git ~dir = { git; dir }
+
+  let create ~git ~dir =
+    let open Fiber.O in
+    Path.mkdir_p dir;
+    let t = load ~git ~dir in
+    let+ () = run t [ "init"; "--bare" ] in
+    t
+  ;;
+
+  let load_or_create ~git ~dir =
+    match Path.exists dir with
+    | true -> load ~git ~dir |> Fiber.return
+    | false -> create ~git ~dir
+  ;;
+
+  module Remote = struct
+    type nonrec t =
+      { repo : t
+      ; handle : string
+      }
+
+    let head_branch = Re.(compile (seq [ str "HEAD branch: "; group (rep1 any); eol ]))
+    let update { repo; handle } = run repo [ "fetch"; handle; "--no-tags" ]
+
+    let default_branch { repo; handle } =
+      let open Fiber.O in
+      let+ contents = run_capture_lines repo [ "remote"; "show"; handle ] in
```

Comment:
`git ls-remote $remote HEAD` is a little more direct.

Reply:
It doesn't do the same though. `ls-remote` will tell me the HEAD revision of the remote *as it is on the remote* at that moment, but looking up the main branch and rev-parsing `remote/branch-name` will tell me what my local remote's HEAD is (no network interaction). So to be on the safe side I'd need to `ls-remote` and then fetch to make sure I have that revision locally.